charms/worker/k8s/src/charm.py

```diff
@@ -1061,14 +1106,62 @@ def _get_external_kubeconfig(self, event: ops.ActionEvent):
         try:
             server = event.params.get("server")
             if not server:
-                log.info("No server requested, use public-address")
-                server = f"{_get_public_address()}:6443"
+                log.info("No server requested, use public address")
+                server = self._get_public_address()
+                if not server:
+                    event.fail("Failed to get public address. Check logs for details.")
+                    return
+                log.info("Found public address: %s", server)
+                port = str(APISERVER_PORT)
+                if self.is_control_plane and self.external_load_balancer.is_available:
+                    log.info("Using external load balancer port as the public port")
+                    port = str(EXTERNAL_LOAD_BALANCER_PORT)
+                server = build_url(server, port, "https")
+                log.info("Formatted server address: %s", server)
             log.info("Requesting kubeconfig for server=%s", server)
             resp = self.api_manager.get_kubeconfig(server)
             event.set_results({"kubeconfig": resp})
         except (InvalidResponseError, K8sdConnectionError) as e:
             event.fail(f"Failed to retrieve kubeconfig: {e}")
 
+    def _get_public_address(self) -> Optional[str]:
+        """Get public address either from external load balancer or from juju.
+
+        Returns:
+            str: public ip address of the unit
+            None: if the public address is not available
+        """
+        if self.is_control_plane and self.external_load_balancer.is_available:
+            log.info("Using external load balancer address as the public address")
+            return self._get_external_load_balancer_address()
+        log.info("Using juju public address as the public address")
+        return _get_juju_public_address()
```

Comment:
maybe a better comment for this function is

>  the "most public" ip address.  But don't ignore the unit's IP in the sans just because there's a load balancer